src/smt/smt_driver.cpp

```diff
@@ -0,0 +1,115 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Andrew Reynolds
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2022 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * The solver for SMT queries in an SolverEngine.
+ */
+
+#include "smt/smt_driver.h"
+
+#include "options/base_options.h"
+#include "options/main_options.h"
+#include "options/smt_options.h"
+#include "prop/prop_engine.h"
+#include "smt/context_manager.h"
+#include "smt/env.h"
+#include "smt/logic_exception.h"
+#include "smt/smt_solver.h"
+
+namespace cvc5::internal {
+namespace smt {
+
+SmtDriver::SmtDriver(Env& env, SmtSolver& smt, ContextManager* ctx)
+    : EnvObj(env), d_smt(smt), d_ctx(ctx)
+{
+}
+
+Result SmtDriver::checkSatisfiability(const std::vector<Node>& assumptions)
+{
+  Assertions& as = d_smt.getAssertions();
+  Result result;
+  try
+  {
+    // then, initialize the assertions
+    as.setAssumptions(assumptions);
+
+    // make the check, where notice smt engine should be fully inited by now
+
+    Trace("smt") << "SmtSolver::check()" << std::endl;
+
+    ResourceManager* rm = d_env.getResourceManager();
+    if (rm->out())
+    {
+      UnknownExplanation why = rm->outOfResources()
+                                   ? UnknownExplanation::RESOURCEOUT
+                                   : UnknownExplanation::TIMEOUT;
+      result = Result(Result::UNKNOWN, why);
+    }
+    else
+    {
+      rm->beginCall();
+
+      bool checkAgain = true;
+      while (checkAgain)
+      {
+        checkAgain = false;
+        // check sat based on the driver strategy
+        result = checkSatNext(checkAgain);
+        // if we were asked to check again
+        if (checkAgain)
+        {
+          Assert(d_ctx != nullptr);
+          as.clearCurrent();
+          d_ctx->notifyResetAssertions();
+          // get the next assertions based on the driver strategy
+          getNextAssertions(as);
+          // finish init to construct new theory/prop engine
+          d_smt.finishInit();
+          // setup
+          d_ctx->setup();
+        }
+      }
+
+      rm->endCall();
```

Comment:
Do we still need to call this even when `rm->out()` is `true`? There is no matching `rm->beginCall()` in that case.

Reply:
I'm not exactly sure.

This code is following the same skeleton as: https://github.com/cvc5/cvc5/blob/main/src/smt/smt_solver.cpp#L136

Let me know if something should change.

Reply:
Oh yeah, I think that it is fine :)